File: messageDotNet/grpc/baseData/Talent.cs
using System;
using UnityEngine;


[Serializable]
public class GrpcTalentExp
{
    public GameMessageCore.TalentType TalentType;
    public long Exp; 
}


[Serializable]
public class GrpcTalentLevel
{
    public GameMessageCore.TalentType TalentType;
    public int Level;
    public int MasterLevel; 
    public GameMessageCore.TalentLevel ToProtoData()
    {
        return new GameMessageCore.TalentLevel()
        {
            Type = TalentType,
            Level = Level,
            MasterLevel = MasterLevel,
        };
    }
}


[Serializable]
public class GrpcTalentNodeData
{
    public int NodeId;
    public int Level; 
    public GameMessageCore.TalentNode ToProtoData()
    {
        return new GameMessageCore.TalentNode()
        {
            NodeId = NodeId,
            Level = Level,
        };
    }
}



[Serializable]
public class GrpcTalentTree
{
    public GameMessageCore.TalentType TalentType;
    public GrpcTalentNodeData[] Nodes; 

    public GameMessageCore.TalentTree ToProtoData()
    {
        GameMessageCore.TalentTree tree = new() { Type = TalentType };
        if (Nodes != null)
        {
            for (int i = 0; i < Nodes.Length; i++)
            {
                tree.UnlockNodes.Add(Nodes[i].ToProtoData());
            }
        }
        return tree;
    }
}


[Serializable]
public class GrpcTalentData
{
    public GrpcTalentLevel[] Levels;
    public GrpcTalentTree[] Trees; 

    public GameMessageCore.TalentData ToProtoData()
    {
        GameMessageCore.TalentData data = new();
        if (Levels != null)
        {
            int lvLength = Levels.Length;
            for (int i = 0; i < lvLength; i++)
            {
                data.LevelData.Add(Levels[i].ToProtoData());
            }
        }
        if (Trees != null)
        {
            int treeLength = Trees.Length;
            for (int i = 0; i < treeLength; i++)
            {
                data.Tree.Add(Trees[i].ToProtoData());
            }
        }
        return data;
    }
}


[Serializable]
public class GrpcTalentTreeUpdate
{
    public GameMessageCore.TalentType TalentType;
    public GrpcTalentNodeData[] AddNodes;
    public GrpcTalentNodeData[] UpdateNodes;
    public int[] RemoveNodes; 

    public GameMessageCore.TalentTreeUpdate ToProtoData()
    {
        GameMessageCore.TalentTreeUpdate data = new() { Type = TalentType };
        if (AddNodes != null)
        {
            int length = AddNodes.Length;
            for (int i = 0; i < length; i++)
            {
                data.AddNodes.Add(AddNodes[i].ToProtoData());
            }
        }
        if (UpdateNodes != null)
        {
            int length = UpdateNodes.Length;
            for (int i = 0; i < length; i++)
            {
                data.UpdateNodes.Add(UpdateNodes[i].ToProtoData());
            }
        }
        if (RemoveNodes != null)
        {
            int length = RemoveNodes.Length;
            for (int i = 0; i < length; i++)
            {
                data.RemoveNodes.Add(RemoveNodes[i]);
            }
        }
        return data;
    }
}

File: messageGo/grpc/baseData/talent.go
package base_data

import "game-message-core/proto"

type GrpcTalentExp struct {
	TalentType proto.TalentType `json:"talentType"`
	Exp        int64            `json:"exp"`
}

func (p *GrpcTalentExp) Clear() {
	p.TalentType = 0
	p.Exp = 0
}

type GrpcTalentLevel struct {
	TalentType  proto.TalentType `json:"talentType"`
	Level       uint32           `json:"level"`
	MasterLevel uint32           `json:"masterLevel"`
}

func (p *GrpcTalentLevel) Clear() {
	p.TalentType = 0
	p.Level = 0
	p.MasterLevel = 0
}

type GrpcTalentNodeData struct {
	NodeId uint32 `json:"nodeId"`
	Level  uint32 `json:"level"`
}

func (p *GrpcTalentNodeData) Clear() {
	p.Level = 0
	p.NodeId = 0
}

type GrpcTalentTree struct {
	TalentType proto.TalentType      `json:"talentType"`
	Nodes      []*GrpcTalentNodeData `json:"nodes"`
}

func (p *GrpcTalentTree) Clear() {
	p.TalentType = 0
	if p.Nodes != nil {
		p.Nodes = p.Nodes[:0]
	}
}

type GrpcTalentData struct {
	Levels []*GrpcTalentLevel `json:"levels"`
	Trees  []*GrpcTalentTree  `json:"trees"`
}

func (p *GrpcTalentData) Clear() {
	if p.Levels != nil {
		p.Levels = p.Levels[:0]
	}
	if p.Trees != nil {
		p.Trees = p.Trees[:0]
	}
}

type GrpcTalentTreeUpdate struct {
	TalentType  proto.TalentType     `json:"talentType"`
	AddNodes    []GrpcTalentNodeData `json:"addNodes"`
	UpdateNodes []GrpcTalentNodeData `json:"updateNodes"`
	RemoveNodes []uint32             `json:"removeNodes"`
}

func (p *GrpcTalentTreeUpdate) Clear() {
	p.TalentType = 0
	if p.AddNodes != nil {
		p.AddNodes = p.AddNodes[:0]
	}
	if p.UpdateNodes != nil {
		p.UpdateNodes = p.UpdateNodes[:0]
	}
	if p.RemoveNodes != nil {
		p.RemoveNodes = p.RemoveNodes[:0]
	}
}
